File: js/ssrfTest.js
const https = require("https");
const http  = require("http");
const { URL } = require("url");

const SSRF_PARAMS = ["url", "uri", "src", "source", "dest", "destination",
                     "redirect", "proxy", "fetch", "load", "link", "host",
                     "callback", "return", "next", "data", "reference"];

const SSRF_TARGETS = [
  { url: "http://169.254.169.254/latest/meta-data/",            label: "AWS Metadata Service",    marker: "ami-id" },
  { url: "http://169.254.169.254/latest/meta-data/iam/",        label: "AWS IAM Credentials",     marker: "security-credentials" },
  { url: "http://metadata.google.internal/computeMetadata/v1/", label: "GCP Metadata Service",    marker: "computeMetadata" },
  { url: "http://localhost/",                                    label: "Localhost Internal",       marker: null },
  { url: "http://127.0.0.1/",                                   label: "Loopback Internal",        marker: null },
  { url: "http://localhost:27017/",                              label: "Local MongoDB",            marker: "mongodb" },
  { url: "http://localhost:6379/",                               label: "Local Redis",              marker: "redis" },
];

async function fetchURL(url, timeout = 5000) {
  return new Promise((resolve) => {
    try {
      const parsed = new URL(url);
      const lib    = parsed.protocol === "https:" ? https : http;
      const req    = lib.get(url, { timeout, rejectUnauthorized: false }, (res) => {
        let data = "";
        res.on("data", (c) => (data += c));
        res.on("end", () => resolve({ status: res.statusCode, body: data }));
      });
      req.on("error",   () => resolve(null));
      req.on("timeout", () => { req.destroy(); resolve(null); });
    } catch { resolve(null); }
  });
}

async function testSSRF(target, siteMap) {
  const results = { name: "Server-Side Request Forgery (SSRF)", passed: true, issues: [] };
  const base    = target.replace(/\/$/, "");

  // Build probe bases — root + all crawled param endpoints
  const probeBases = new Set([base]);
  if (siteMap?.paramUrls) {
    for (const { url } of siteMap.paramUrls) {
      try { probeBases.add(new URL(url).origin + new URL(url).pathname); }
      catch { }
    }
  }

  for (const probeBase of probeBases) {
    for (const param of SSRF_PARAMS) {
      for (const { url: ssrfUrl, label, marker } of SSRF_TARGETS) {
        const testURL = `${probeBase}?${param}=${encodeURIComponent(ssrfUrl)}`;
        const res     = await fetchURL(testURL);
        if (!res || res.status === 404) continue;

        const bodyLower = res.body.toLowerCase();
        const triggered = marker
          ? bodyLower.includes(marker.toLowerCase())
          : (res.status === 200 && res.body.length > 100 &&
             (bodyLower.includes("localhost") || bodyLower.includes("html")));

        if (triggered) {
          results.passed = false;
          results.issues.push({
            title:    `SSRF Vulnerability — ${label} Accessible`,
            location: `Parameter "?${param}=" at ${probeBase}`,
            detail:   `Server fetched internal URL "${ssrfUrl}" and returned its contents (HTTP ${res.status}).`,
            severity: label.includes("AWS") || label.includes("GCP") ? "CRITICAL" : "HIGH",
            risk:     "Attackers trick your server into fetching internal systems — stealing cloud credentials or mapping your private network.",
            fixes: [
              {
                label: "Validate URLs against a strict allowlist",
                code:  "const ALLOWED_HOSTS = ['api.yourservice.com'];\n\nfunction isSafeURL(inputUrl) {\n  try {\n    const parsed = new URL(inputUrl);\n    const blocked = ['localhost','127.0.0.1','0.0.0.0','169.254.169.254'];\n    if (blocked.some(b => parsed.hostname.startsWith(b))) return false;\n    return ALLOWED_HOSTS.includes(parsed.hostname);\n  } catch { return false; }\n}"
              },
              {
                label: "Block metadata IP at network level",
                code:  "iptables -A OUTPUT -d 169.254.169.254 -j DROP"
              }
            ]
          });
          break;
        }
      }
    }
  }

  return results;
}

module.exports = testSSRF;

File: C#/SsrfTest.cs
namespace SecurityTester.Tests;

public class SsrfTest
{
    private static readonly string[] SsrfParams =
        { "url", "uri", "src", "source", "dest", "destination",
          "redirect", "proxy", "fetch", "load", "link", "host", "callback" };

    private record SsrfTarget(string Url, string Label, string? Marker);
    private static readonly List<SsrfTarget> Targets = new()
    {
        new("http://169.254.169.254/latest/meta-data/", "AWS Metadata Service",    "ami-id"),
        new("http://169.254.169.254/latest/meta-data/iam/", "AWS IAM Credentials","security-credentials"),
        new("http://localhost/",                         "Localhost Internal",      null),
        new("http://127.0.0.1/",                         "Loopback Internal",       null),
        new("http://localhost:27017/",                   "Local MongoDB",           "mongodb"),
        new("http://localhost:6379/",                    "Local Redis",             "redis"),
    };

    public static async Task<TestResult> RunAsync(string target, HttpClient client, SiteMap? siteMap = null)
    {
        var result  = new TestResult { Name = "Server-Side Request Forgery (SSRF)" };
        var baseUrl = target.TrimEnd('/');

        // Build probe bases — root + all crawled param endpoints
        var probeBases = new HashSet<string>(StringComparer.OrdinalIgnoreCase) { baseUrl };
        if (siteMap?.ParamUrls != null)
        {
            foreach (var p in siteMap.ParamUrls)
            {
                try
                {
                    var u = new Uri(p.Url);
                    probeBases.Add($"{u.Scheme}://{u.Host}{u.AbsolutePath}");
                }
                catch { }
            }
        }

        foreach (var probeBase in probeBases)
        foreach (var param in SsrfParams)
        {
            foreach (var ssrfTarget in Targets)
            {
                var testUrl = $"{baseUrl}?{param}={Uri.EscapeDataString(ssrfTarget.Url)}";
                try
                {
                    var res    = await client.GetAsync(testUrl);
                    var body   = (await res.Content.ReadAsStringAsync()).ToLower();
                    int status = (int)res.StatusCode;
                    if (status == 404) continue;

                    bool triggered = ssrfTarget.Marker != null
                        ? body.Contains(ssrfTarget.Marker.ToLower())
                        : (status == 200 && body.Length > 100 &&
                           (body.Contains("localhost") || body.Contains("html") || body.Contains("server")));

                    if (triggered)
                    {
                        result.Passed = false;
                        result.Issues.Add(new SecurityIssue
                        {
                            Title    = $"SSRF Vulnerability — {ssrfTarget.Label} Accessible",
                            Location = $"Parameter \"?{param}=\" at {baseUrl}",
                            Detail   = $"Server fetched internal URL \"{ssrfTarget.Url}\" and returned contents (HTTP {status}).",
                            Severity = ssrfTarget.Label.Contains("AWS") ? "CRITICAL" : "HIGH",
                            Risk     = "Attackers trick your server into fetching internal systems — stealing cloud credentials or mapping your private network.",
                            Fixes    = new()
                            {
                                new() { Label = "C# — validate URLs against strict allowlist",
                                        Code  = "private static readonly string[] AllowedHosts = { \"api.service.com\", \"cdn.yoursite.com\" };\n\nbool IsSafeUrl(string inputUrl)\n{\n    if (!Uri.TryCreate(inputUrl, UriKind.Absolute, out var uri)) return false;\n    var blocked = new[] { \"localhost\", \"127.0.0.1\", \"0.0.0.0\", \"169.254.169.254\" };\n    if (blocked.Any(b => uri.Host.StartsWith(b))) return false;\n    return AllowedHosts.Contains(uri.Host);\n}" },
                                new() { Label = "Block metadata IP at network/firewall level",
                                        Code  = "# Linux iptables:\niptables -A OUTPUT -d 169.254.169.254 -j DROP\niptables -A OUTPUT -d 169.254.0.0/16  -j DROP" }
                            }
                        });
                        goto nextParam;
                    }
                }
                catch { }
            }
            nextParam:;
        }

        return result;
    }
}
